File: app/src/main/java/net/roningrum/barvolume/MainActivity.java
package net.roningrum.barvolume;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    private EditText edtWidht;
    private EditText edtLenght;
    private EditText edtHeight;
    private Button btnCalculate;
    private TextView tvResult;

    private static final String STATE_RESULT = "state_result";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        edtWidht = findViewById(R.id.edt_width);
        edtLenght = findViewById(R.id.edt_length);
        edtHeight = findViewById(R.id.edt_height);
        btnCalculate = findViewById(R.id.btn_calculate);
        tvResult = findViewById(R.id.tv_result);

        btnCalculate.setOnClickListener(this);

        if(savedInstanceState !=null){
            String result = savedInstanceState.getString(STATE_RESULT);
            tvResult.setText(result);
        }
    }

    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.btn_calculate) {
            String inputLength = edtLenght.getText().toString();
            String inputWidht = edtWidht.getText().toString();
            String inputHeight = edtHeight.getText().toString();

            boolean isEmptyFields = false;
            boolean isInvalidDouble = false;

            if (TextUtils.isEmpty(inputLength)) {
                isEmptyFields = true;
                edtLenght.setError("Field ini tidak boleng kosong");
            }
            if (TextUtils.isEmpty(inputWidht)) {
                isEmptyFields = true;
                edtWidht.setError("Field ini tidak boleng kosong");
            }
            if (TextUtils.isEmpty(inputHeight)) {
                isEmptyFields = true;
                edtHeight.setError("Field ini tidak boleng kosong");
            }
            Double length = toDouble(inputLength);
            Double widht = toDouble(inputWidht);
            Double height = toDouble(inputHeight);

            if (length == null) {
                isInvalidDouble = true;
                edtLenght.setError("Field ini harus berupa angka yang valid");
            }
            if (widht == null) {
                isInvalidDouble = true;
                edtWidht.setError("Field ini harus berupa angka yang valid");
            }
            if (height == null) {
                isInvalidDouble = true;
                edtHeight.setError("Field ini harus berupa angka yang valid");
            }
            if (!isEmptyFields && !isInvalidDouble) {
                double volume = length * widht * height;
                tvResult.setText(String.valueOf(volume));
            }

        }
    }

    private Double toDouble(String str) {
        try{
            return  Double.valueOf(str);
        } catch (NumberFormatException e){
            return  null;
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(STATE_RESULT, tvResult.getText().toString());
    }
}

File: barvolumekotlinverse/src/main/java/net/roningrum/barvolumekotlinverse/MainActivity.kt
package net.roningrum.barvolumekotlinverse

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import java.lang.NumberFormatException

class MainActivity : AppCompatActivity(), View.OnClickListener {
    companion object{
        private const val STATE_RESULT = "state_result"
    }
    private lateinit var edtWidht: EditText
    private lateinit var edtLenght: EditText
    private lateinit var edtHeight: EditText
    private lateinit var btnCalculate: Button
    private lateinit var tvResult: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        edtWidht = findViewById(R.id.edt_width)
        edtLenght = findViewById(R.id.edt_length)
        edtHeight = findViewById(R.id.edt_height)
        btnCalculate = findViewById(R.id.btn_calculate)
        tvResult = findViewById(R.id.tv_result)

        btnCalculate.setOnClickListener(this)

        if(savedInstanceState != null){
            val result = savedInstanceState.getString(STATE_RESULT) as String
            tvResult.text = result
        }
    }

    override fun onClick(v: View) {
        if (v.id == R.id.btn_calculate) {
            val inputLenght = edtLenght.text.toString().trim()
            val inputWidht = edtWidht.text.toString().trim()
            val inputHeight = edtHeight.text.toString().trim()

            var isEmptyFields = false
            var isInvalidDouble = false

            if (TextUtils.isEmpty(inputLenght)) {
                isEmptyFields = true
                edtLenght.error = "Field ii tidak boleh kosong"
            }

            if (TextUtils.isEmpty(inputWidht)) {
                isEmptyFields = true
                edtWidht.error = "Field ii tidak boleh kosong"
            }
            if (TextUtils.isEmpty(inputHeight)) {
                isEmptyFields = true
                edtHeight.error = "Field ii tidak boleh kosong"
            }
            val length = toDouble(inputLenght)
            val widht = toDouble(inputWidht)
            val height = toDouble(inputHeight)

            if(length == null){
                isInvalidDouble = true
                edtLenght.error = "Field ini harus berupa angka yang valid"
            }
            if(widht == null){
                isInvalidDouble = true
                edtWidht.error = "Field ini harus berupa angka yang valid"
            }
            if(height == null){
                isInvalidDouble = true
                edtHeight.error = "Field ini harus berupa angka yang valid"
            }
            if(!isEmptyFields && !isInvalidDouble){
                val volume = length as Double * widht as Double * height as Double
                tvResult.text = volume.toString()
            }

        }

    }

    private fun toDouble(str: String): Double? {
        return try{
            str.toDouble()
        } catch(e: NumberFormatException){
            null
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(STATE_RESULT, tvResult.text.toString())
    }
}
